File: hadoop/average-temperature-java/src/test/java/name/ichimonji10/average_temperature/TemperatureAveragerTest.java
package name.ichimonji10.average_temperature;

import java.io.IOException;
import java.util.Arrays;

import org.apache.hadoop.io.FloatWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Reducer;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

/**
 * Tests for {@link TemperatureAverager}.
 */
class TemperatureAveragerTest {

    /** Test {@link TemperatureAverager#reduce}. */
    @Test
    void testReduce() throws IOException, InterruptedException {
        final Text key = new Text(new String("foo"));
        final Iterable<FloatWritable> values = Arrays.asList(
            new FloatWritable(new Float(1.2)),
            new FloatWritable(new Float(3.4))
        );
        final Reducer.Context context = Mockito.mock(Reducer.Context.class);
        final TemperatureAverager temperatureAverager = new TemperatureAverager();
        temperatureAverager.reduce(key, values, context);
        Mockito.verify(context).write(
            new Text(new String("foo")),
            new FloatWritable((new Float(1.2) + new Float(3.4)) / 2)
        );
    }
}

File: hadoop/average-temperature-scala/src/test/java/name/ichimonji10/average_temperature/TemperatureAveragerTest.scala
package name.ichimonji10.average_temperature

import java.lang.Float
import org.apache.hadoop.io.FloatWritable
import org.apache.hadoop.io.Text
import org.apache.hadoop.mapreduce.Reducer
import org.mockito.Mockito
import org.scalatest.FlatSpec
import scala.collection.JavaConverters._

class TemperatureAveragerTest extends FlatSpec {
  private val rawValues: Iterable[Float] = Seq(new Float(1.2), new Float(3.4))
  private val rawAverage: Float = new Float(rawValues.reduce(_ + _) / new Float(2))

  behavior of "reduce function"

  it should "average the given measurements" in {
    val key: Text = new Text(new String("foo"))
    val values: java.lang.Iterable[FloatWritable] = rawValues.map(new FloatWritable(_)).asJava
    val context: Reducer[Text, FloatWritable, Text, FloatWritable]#Context =
      Mockito.mock(
        classOf[Reducer[Text, FloatWritable, Text, FloatWritable]#Context]
      )
    new TemperatureAverager().reduce(key, values, context)
    Mockito.verify(context).write(key, new FloatWritable(rawAverage))
  }
}
